src/YaccConstructor/Program.fs

```diff
@@ -88,37 +114,47 @@ let () =
             Some tmpName
         else None
 
+    let generatorNameD = 
+            match !generatorName with
+            |Some x->x
+            |None->""
+
     let generateSomething = ref true
 
     let printItems iName items deft =
-        fun _ ->
             generateSomething := false
             printfn "\nAvailable %s: " iName
             Seq.map (fun x -> x + (if Some x = deft then " (default)" else "")) items
             |> String.concat "\n    "
             |> fun x -> printf "    %s\n" x
 
-    let commandLineSpecs =
-        ["-f", ArgType.String (fun s -> feName := Some s), "Frontend name. Use -af to list available."
-         "-af", ArgType.Unit (printItems "frontends" addinFrontendNames !feName), "Available frontends"
-         "-g", ArgType.String 
-            (fun s -> 
-                match Array.toList (s.Split ' ') with
+    let argv = System.Environment.GetCommandLineArgs()
+    let parser = ArgumentParser.Create<CLIArguments>(errorHandler = ProcessExiter())
+    let args = parser.Parse argv.[1..]
+    feName := args.GetResult (<@ F @>, defaultValue = !feName)
+    let genString= args.GetResult (<@ G @>, defaultValue = generatorNameD)
+
+    match Array.toList (genString.Split ' ') with
                 | name::[] -> generatorName := Some name; generatorParams := None
                 | name::parameters -> generatorName := Some name; generatorParams := Some (String.concat " " parameters)
                 | _ -> failwith "You need to specify generator name"
-            ), "Generator name. Use -ag to list available."
-         "-ag", ArgType.Unit (printItems "generators" addinGeneratorNames !generatorName), "Available generators"
-         "-c", ArgType.String (fun s -> conversions.Add s), "Conversion applied in order. Use -ac to list available."
-         "-ac", ArgType.Unit (printItems "conversions" addinConversionNames None), "Available conversions"
-         "-D", ArgType.String (fun s -> userDefs := !userDefs @ [s]), "User defined constants for YardFrontend lexer."
-         "-U", ArgType.String (fun s -> userDefs := List.filter ((<>) s) !userDefs), 
-                "Remove previously defined constants for YardFrontend lexer."
-         "-i", ArgType.String (fun s ->
-                                   testFile := System.IO.Path.GetFileName s |> Some
-                                   testsPath := System.IO.Path.GetDirectoryName s |> Some), "Input grammar"         
-         ] |> List.map (fun (shortcut, argtype, description) -> ArgInfo(shortcut, argtype, description))
-    ArgParser.Parse commandLineSpecs
+
+    for c in args.GetResults <@ C @> do
```

Comment:
This long list of loops and conditions can be replaced with pattern-matching while iterating over all arguments.